sdks/java/harness/src/main/java/org/apache/beam/fn/harness/control/ProcessBundleHandler.java

```diff
@@ -491,14 +514,25 @@ private BundleProcessor createBundleProcessor(
       }
     }
 
-    // Instantiate a State API call handler depending on whether a State ApiServiceDescriptor
-    // was specified.
-    HandleStateCallsForBundle beamFnStateClient =
-        bundleDescriptor.hasStateApiServiceDescriptor()
-            ? new BlockTillStateCallsFinish(
-                beamFnStateGrpcClientCache.forApiServiceDescriptor(
-                    bundleDescriptor.getStateApiServiceDescriptor()))
-            : new FailAllStateCallsForBundle(processBundleRequest);
+    // Instantiate a State API call handler depending on whether a State ApiServiceDescriptor was
+    // specified. If pipeline is batch, use a CachingBeamFnStateClient to store state responses.
+    // User state caching is currently not supported in streaming mode.
+    HandleStateCallsForBundle beamFnStateClient;
+    if (bundleDescriptor.hasStateApiServiceDescriptor()) {
```

Comment:
Can you explain a bit more why `hasStateApiServiceDescriptor` is true = batch?

Reply:
`hasStateApiServiceDescriptor` determines if a state handler is used at all, and `options.as(StreamingOptions.class).isStreaming()` determines if it is batch or streaming. Updated comments to show this more clearly

Reply:
O I see. `options.as(StreamingOptions.class).isStreaming()` is used at line 530.